### PR TITLE
4764: Related materials

### PR DESCRIPTION
**Pull request by kasperg**
#### Link to issue

https://platform.dandigbib.org/issues/4764

#### Description

This PR adds a new React app displaying related materials for an object on the object page.

#### Screenshot of the result

Ting object | Harry Potter og fangen fra Azkaban | Ding2 2020-05-25 00-03-23

#### Checklist

- [x] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [x] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [x] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

To make this work for anonymous users were we need to access OpenPlatform without a token for an authenticated user we pull in support for agency tokens for #1617. Note that if that PR is updated and merged this will likely result in conflicts for us as well.

Scrutinizer comments are either "imported" through 6f1f54b or contradictory.

### CODE REVIEW

File: modules/ding_react/plugins/content_types/related_materials.inc [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    $data = [
+      'subjects' => ding_react_related_materials_format_strings($object->getSubjects()),
+      'categories' => ding_react_related_materials_format_strings($object->getAudience()),
+      'sources' => ding_react_related_materials_format_strings($sources),
+      'exclude-title' => ding_react_related_materials_format_string($object->getTitle()),
+      // We cannot use url() here as it will encode the colon in the placeholder.
+      'search-url' => '/search/ting/:query?sort=:sort',
+      'material-url' => '/ting/object/:pid',
+      'cover-service-url' => 'https://cover.dandigbib.org/api',
```
Comment (cableman) [inline modules/ding_react/plugins/content_types/related_materials.inc:65]: Why is the cover end-point hardcoded into this and not fetched from configuration in the CMS backend.

Comment (kasperg) [reply modules/ding_react/plugins/content_types/related_materials.inc:65]: Configuration for this has been added in #1640. The two PRs are somewhat intertwined at the moment 😕 .

File: modules/ding_react/ding_react.module [hunk 2]
```diff
@@ -20,9 +20,10 @@ function ding_react_libraries_info() {
   return [
     'ddb-react' => [
       'name' => 'DDB React',
-      'vendor url' => 'https://github.com/reload/ddb-react',
-      'download url' => 'https://github.com/reload/ddb-react/releases/download/latest/dist.zip',
-      'version' => '1.0.0',
+      'vendor url' => 'https://github.com/danskernesdigitalebibliotek/ddb-react',
+      'download url' => 'https://github.com/danskernesdigitalebibliotek/ddb-react/releases/download/latest/dist.zip',
+      // We have to set the version to something for Library API to work.
+      'version' => 'latest',
```
[the remark this answers is not in the document]
Comment (kasperg) [reply modules/ding_react/ding_react.module]: It is currently versioned using the same approach as some of our other libraries use where [we track `master` on `master` but set tags for releases](https://github.com/ding2/ding2/commit/b0e809ef5ee93e2c53e3e281fdfb7a2a793456f6#diff-12abfbcee6201316ed4b595cdbb9f77aL630).

As such the version set by Libraries API does not really make a difference. However in 
d2445b8 I have updated the module to extract the version used from the library so you can at least see it by running `drush libraries-list`.

File: modules/ding_react/plugins/content_types/related_materials.inc [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  $ting_objects = array_map(function (ctools_context $context) {
+    $data = $context->data;
+    if ($data instanceof TingCollection) {
+      return $data->getPrimary_object()->getTingObject();
+    }
+    elseif ($data instanceof TingEntity) {
+      return $data->getTingObject();
+    }
+  }, $contexts_with_data);
+  $object = array_shift($ting_objects);
```
[the remark this answers is not in the document]
Comment (kasperg) [reply modules/ding_react/plugins/content_types/related_materials.inc:42]: I have tried to add an explanatory code comment in 826f8e9.

File: modules/ding_react/plugins/content_types/related_materials.inc [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    // list is not sufficient. NB: There is currently no admin UI for setting
+    // this.
+    $sources = variable_get('ding_react_related_materials_sources', implode(',', [
+      'bibliotekskatalog',
+      'ereolen',
+      'ereolen global',
+      'comics plus',
+      'ebook central',
+      'rbdigital magazines',
+    ]));
```
[the remark this answers is not in the document]
Comment (kasperg) [reply modules/ding_react/plugins/content_types/related_materials.inc:57]: This list has been defined by the business in the issue. I have tried to add a code comment explaining the reasoning as I recall it in 826f8e9.

**Comment by kasperg** [comment]
@cableman I have intentionally merged in all of #1617. Otherwise this cannot be tested locally. 

Once #1617 is merged the commits are no longer relevant in this PR. We can then remove the merge commit by hand but I think Git is clever enough to handle it for us.

**Comment by cableman** [comment]
Approved with the condition that #1617 gets removed from this PR before merge :-)

**Pull request merged**